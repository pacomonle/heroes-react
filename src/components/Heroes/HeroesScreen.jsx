import React, { useMemo } from 'react'
import { Redirect, useParams } from 'react-router-dom'
import GetHeroById from '../helpers/GetHeroById';
import PropTypes from 'prop-types'

const HeroesScreen = ({history}) => {
// recuperar los params que viene por url
   const params = useParams()
   const {heroeId} = params
   
   // optimizar - evitar renderizacions 
   const hero = useMemo(() => GetHeroById(heroeId), [heroeId])
  // const hero = GetHeroById(heroeId)
   // console.log(hero)
   if(!hero){ return <Redirect to='/' /> }

   const { 
    superhero,
    alter_ego,
    first_appearance,
    publisher,
    characters
          } = hero


    const handleReturn = () =>{
  // validar longitud del history
        if(history.length <= 2){
            history.push('/')
        }else{
               // regresar atras
            history.goBack()
        }

      
    }

    return (
        <div className='row mt-4'>
           <div className='col-4'>
               <img src={`../assets/heroes/${heroeId}.jpg`} alt={superhero} 
               className='img-thumbnail animate__animated animate__fadeInLeft'/>
           </div>
           <div className='col-8 animate__animated animate__fadeIn'>
               <h3>{superhero}</h3>
               <ul className='list-group list-group-flush'>
                    <li className='list-group-item'>
                      <strong>ALTER EGO: </strong> {alter_ego}
                    </li>
                    <li className='list-group-item'>
                      <strong>PUBLISHER: </strong> {publisher}
                    </li>
                    <li className='list-group-item'>
                      <strong>FIRST APPEARENCE: </strong> {first_appearance}
                    </li>
               </ul>
               <h5>Characters</h5>
               <p>{characters}</p>
               <button 
               className='btn btn-outline-info'
               onClick={handleReturn}
               type="button">
               Return
               </button>
           </div>
        </div>
    )
}

HeroesScreen.propTypes = {
    history : PropTypes.object.isRequired
 }

export default HeroesScreen
