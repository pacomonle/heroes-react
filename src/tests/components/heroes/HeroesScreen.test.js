import React from 'react';
import '@testing-library/jest-dom'
import HeroesScreen from '../../../components/Heroes/HeroesScreen';
import { MemoryRouter, Redirect, Route } from 'react-router-dom';
const { mount } = require("enzyme")

describe('Pruebas en el HeroesScreen', () => {
   // para mockear(simular el history)
    const historyMock={
        push: jest.fn(),
        length: 1,
        goBack: jest.fn()
     }


   
    test('debe de mostrar el componente redirect si no hay argumentos en el URL', () => {

        const wrapper = mount(
            // pasar por props la ruta - url de entrada
            <MemoryRouter initialEntries={ ['/hero'] }>
                 <HeroesScreen history={historyMock} />
            </MemoryRouter>
            
            )

        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find(Redirect).exists()).toBe(true)
    })
    

    test('debe de mostrar un hero si el parÃ¡metro existe y se encuentra', () => {
        

        const wrapper = mount(
                 // pasar por props la ruta - url de entrada ... heroe/heroeID
            <MemoryRouter initialEntries={['/heroe/marvel-spider']}>
                <Route path="/heroe/:heroeId" component={ HeroesScreen } />
            </MemoryRouter>
        );
       // console.log(wrapper.html())
        expect( wrapper.find('.row').exists() ).toBe(true);
        

    });

    test('debe de regresar a la pantalla anterior con PUSH', () => {
        const history = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn(),
        }

     
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-spider']}>
                <Route 
                    path="/heroe/:heroeId" 
                    component={ () => <HeroesScreen history={ history } /> }
                />
            </MemoryRouter>
        );
        
        wrapper.find('button').prop('onClick')();

        // console.log(history.length)
        
        expect( history.push ).toHaveBeenCalledWith('/');
        expect( history.goBack ).not.toHaveBeenCalled();


    });

    test('debe de regresar a la pantalla anterior GOBACK', () => {
       
        const history = {
            length: 10,
            push: jest.fn(),
            goBack: jest.fn(),
        }
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/marvel-spider']}>
                <Route 
                    path="/heroe/:heroeId" 
                    component={ () => <HeroesScreen history={ history } /> }
                />
            </MemoryRouter>
        );
        
        wrapper.find('button').prop('onClick')();
        
        expect( history.push ).toHaveBeenCalledTimes(0);
        expect( history.goBack ).toHaveBeenCalled();

    })

    test('debe de llamar el redirect si el hero no existe', () => {

        const history = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn(),
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider123123123']}>
                <Route 
                    path="/heroe/:heroeId" 
                    component={ () => <HeroesScreen history={ history } /> }
                />
            </MemoryRouter>
        );
        
 // al hacer el redirect el componente HeroesScreen no existe, es un string vacio '' - ver snapShot 
        expect( wrapper.text() ).toBe('');
        
    })
  
    

})
