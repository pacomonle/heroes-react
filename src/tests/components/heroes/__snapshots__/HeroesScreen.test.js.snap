// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el HeroesScreen debe de mostrar el componente redirect si no hay argumentos en el URL 1`] = `""`;
